game/src/main/kotlin/content/entity/world/music/Music.kts

```diff
@@ -62,11 +92,128 @@ move({ !it.isBot }) { player ->
 interfaceOption("Play", "tracks", "music_player") {
     val index = itemSlot / 2
     if (player.hasUnlocked(index)) {
-        player["playing_song"] = true
         player.playTrack(index)
     }
 }
 
+interfaceOption("Play", "playlist", "music_player") {
+    val index = player["playlist_slot_${itemSlot+1}", 32767]
+    if (player.hasUnlocked(index)) {
+        player.playTrack(index)
+    }
+}
+
+interfaceOption("Add to playlist", "tracks", "music_player") {
+    player.addToPlaylist(itemSlot)
+}
+
+interfaceOption("Remove from playlist", id = "music_player") {
+    player.removeSongFromPlaylist(itemSlot, component == "tracks")
+}
+
+interfaceOption("Playlist on/off", "playlist_toggle", "music_player") {
+    player.togglePlaylist()
+}
+
+interfaceOption("Clear Playlist", "clear_playlist", "music_player") {
+    player.clearPlaylist()
+}
+
+interfaceOption("Shuffle on/off", "shuffle_playlist", "music_player") {
+    player.togglePlaylistShuffle()
+}
+
+interfaceSwap(fromId = "music_player", fromComponent = "playlist") { player ->
+    val fromSong = player["playlist_slot_${fromSlot+1}", 32767]
+    val toSong = player["playlist_slot_${toSlot+1}", 32767]
+
+    player["playlist_slot_${fromSlot+1}"] = toSong
+    player["playlist_slot_${toSlot+1}"] = fromSong
+}
+
+songEnd {player ->
+    player["playing_song"] = false
+    if (player["playlist_enabled", false]) {
+        playNextPlaylistTrack(player, songIndex)
+        return@songEnd
+    }
```

Comment:
```suggestion
    if (player["playlist_enabled", false] && playNextPlaylistTrack(player, songIndex)) {
        return@songEnd
    }
```

If a playlist is cleared `playlist_enabled` will still be true but there won't be any tracks, so it would be a good idea to default to playing the area track (or alternatively setting playlist_enabled = false when clearing a playlist?

Reply:
Good catch. Thanks, resolved